server/src/main/java/io/deephaven/server/notebook/FilesystemStorageServiceGrpcImpl.java

```diff
@@ -0,0 +1,210 @@
+import io.deephaven.proto.backplane.grpc.StorageServiceGrpc;
+import io.deephaven.server.session.SessionService;
+import io.grpc.stub.StreamObserver;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Optional;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Stream;
+
+@Singleton
+public class FilesystemStorageServiceGrpcImpl extends StorageServiceGrpc.StorageServiceImplBase {
+    private static final Logger log = LoggerFactory.getLogger(FilesystemStorageServiceGrpcImpl.class);
+
+    private static final String STORAGE_PATH =
+            Configuration.getInstance().getStringWithDefault("storage.path", "<workspace>/storage")
+                    .replace("<workspace>", Configuration.getInstance().getWorkspacePath());
+
+    private final Path root = Paths.get(STORAGE_PATH).normalize();
+    private final SessionService sessionService;
+
+    @Inject
+    public FilesystemStorageServiceGrpcImpl(SessionService sessionService) {
+        this.sessionService = sessionService;
+        try {
+            Files.createDirectories(root);
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to initialize storage", e);
+        }
+    }
+
+    private Path resolveOrThrow(String relativePath) {
+        Path resolved = root.resolve(relativePath).normalize();
+        if (resolved.startsWith(root)) {
+            return resolved;
+        }
+        throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Invalid path: " + relativePath);
+    }
+
+    @Override
+    public void listItems(ListItemsRequest request, StreamObserver<ListItemsResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            sessionService.getCurrentSession();
+
+            ListItemsResponse.Builder builder = ListItemsResponse.newBuilder();
+            PathMatcher matcher = request.hasFilterGlob() ? makeMatcher(request.getFilterGlob()) : ignore -> true;
+            Path dir = resolveOrThrow(request.getPath());
+            try (Stream<Path> list = Files.list(dir)) {
+                for (Path p : (Iterable<Path>) list::iterator) {
+                    if (!matcher.matches(dir.relativize(p))) {
+                        continue;
+                    }
+                    boolean isDirectory = Files.isDirectory(p);
+                    builder.addItems(FileInfo.newBuilder()
+                            .setPath(p.getFileName().toString())
+                            .setSize(isDirectory ? 0 : Files.size(p))
+                            .setKind(isDirectory ? FileKind.DIRECTORY : FileKind.FILE)
```

Comment:
My initial thinking here was that the clients should (at least to start with) be agnostic of the implementation - if the admin decides to move a directory and link it with a symlink, the client shouldn't have to work out how to ask "okay, this is a symlink, what is the type of the target, resolve the target", but should be able to just treat it like a "file" or "directory".

That said, I left this as an enum so that we could add other file types like that, but wanted to be sure we had a nice way to convey all of their metadata... without attempting to design a general purpose filesystem api.